File: ml/lucent_ml/pipeline/segment.py
"""Stage 2 — group words into sentences that carry page + bbox geometry.

We reconstruct text per page from the ordered word list, run syntok sentence
segmentation, then map each sentence's character span back to the words it
covers so every sentence keeps real geometry. Sentences never span pages
because we segment one page at a time.
"""
from dataclasses import dataclass
from syntok.segmenter import process as syntok_process

from .parse import Word


@dataclass(frozen=True)
class Sentence:
    text: str
    page: int
    char_start: int                                   # offset within the page's reconstructed text
    char_end: int
    word_bboxes: list[tuple[float, float, float, float]]


def _page_text_and_spans(words: list[Word]) -> tuple[str, list[tuple[int, int, Word]]]:
    """Join words with single spaces; record each word's [start,end) char span."""
    parts: list[str] = []
    spans: list[tuple[int, int, Word]] = []
    cursor = 0
    for w in words:
        if parts:
            cursor += 1  # the space separator
        start = cursor
        parts.append(w.text)
        cursor += len(w.text)
        spans.append((start, cursor, w))
    return (" ".join(parts), spans)


def segment(words: list[Word]) -> list[Sentence]:
    sentences: list[Sentence] = []
    pages = sorted({w.page for w in words})
    for page in pages:
        page_words = [w for w in words if w.page == page]
        text, spans = _page_text_and_spans(page_words)

        # syntok gives token-level offsets; we use its sentence grouping and
        # re-locate each sentence in `text` via running search to get char spans.
        search_from = 0
        for paragraph in syntok_process(text):
            for sent in paragraph:
                raw = "".join(t.spacing + t.value for t in sent).strip()
                if not raw:
                    continue
                idx = text.find(raw, search_from)
                if idx == -1:
                    idx = text.find(raw)
                if idx == -1:
                    continue
                start, end = idx, idx + len(raw)
                search_from = end
                bboxes = [w.bbox for (s0, s1, w) in spans if s0 < end and s1 > start]
                if not bboxes:
                    continue
                sentences.append(Sentence(
                    text=raw, page=page, char_start=start, char_end=end, word_bboxes=bboxes,
                ))
    return sentences

File: web/lib/summarize/segment.ts
/** Stage: group extracted PDF words into sentences that carry page + bbox
 * geometry -- the browser-side mirror of `ml/lucent_ml/pipeline/segment.py`.
 *
 * Words are reconstructed into a single string per page (so a sentence never
 * spans two pages), segmented with the browser's built-in `Intl.Segmenter`,
 * and each sentence's character span is mapped back to the words it covers so
 * every sentence keeps real geometry. */
import type { ExtractedWord, Sentence } from "./types";

interface WordSpan {
  start: number;
  end: number;
  word: ExtractedWord;
}

function pageTextAndSpans(words: ExtractedWord[]): { text: string; spans: WordSpan[] } {
  const parts: string[] = [];
  const spans: WordSpan[] = [];
  let cursor = 0;
  for (const w of words) {
    if (parts.length > 0) cursor += 1; // the space separator
    const start = cursor;
    parts.push(w.text);
    cursor += w.text.length;
    spans.push({ start, end: cursor, word: w });
  }
  return { text: parts.join(" "), spans };
}

export function segmentIntoSentences(words: ExtractedWord[]): Sentence[] {
  const sentences: Sentence[] = [];
  const pages = Array.from(new Set(words.map((w) => w.page))).sort((a, b) => a - b);
  const segmenter = new Intl.Segmenter(undefined, { granularity: "sentence" });

  for (const page of pages) {
    const pageWords = words.filter((w) => w.page === page);
    const { text, spans } = pageTextAndSpans(pageWords);
    if (!text.trim()) continue;

    for (const { segment, index } of segmenter.segment(text)) {
      const leadingWs = segment.length - segment.trimStart().length;
      const raw = segment.trim();
      if (!raw) continue;
      const start = index + leadingWs;
      const end = start + raw.length;
      const bboxes = spans
        .filter((s) => s.start < end && s.end > start)
        .map((s) => s.word.bbox);
      if (bboxes.length === 0) continue;
      sentences.push({ text: raw, page, bboxes });
    }
  }
  return sentences;
}
